File: qbopt/frontend/raising_division.py
"""Recognize a sign-extended word dividend as one scalar signed division."""

from dataclasses import replace

from qbopt.model import ir
from qbopt.model import mir


def scalar(body: mir.MirBody) -> mir.MirBody:
    definitions = {value: op for block in body.blocks for op in block.ops for value in op.defines}

    def raised(op: mir.Op) -> mir.Op:
        if op.kind is not mir.Kind.DIV or op.name != "idiv" or len(op.args) != 3 or len(op.results) != 2:
            return op
        high, low, divisor = op.args
        if not all(isinstance(arg, mir.Held) and arg.width == 2 for arg in (*op.args, *op.results)):
            return op
        if op.loads or op.stores or not _extends(high, low, definitions):
            return op
        # `idiv r16` ties dx:ax to the pair it writes. Dropping the high half
        # drops its tie with it: the lowering builds a fresh high from `cwd`,
        # so nothing carries the old dx into the remainder any more. A tie to
        # anything else is a shape this does not know.
        if not op.merges.keys() <= {high.value, low.value}:
            return op
        remaining = {low.value, divisor.value}
        return replace(
            op,
            kind=mir.Kind.DIVMOD,
            args=(low, divisor),
            merges={was: now for was, now in op.merges.items() if was != high.value},
            uses=tuple(value for value in op.uses if value != high.value or value in remaining),
        )

    return replace(body, blocks=tuple(replace(block, ops=tuple(map(raised, block.ops))) for block in body.blocks))


def _extends(high: mir.Held, low: mir.Held, definitions: dict) -> bool:
    """Whether `high` is the top word of the sign extension of `low`.

    `cwd` is BC's way of saying it, but not the only one: a dividend that
    reached a long by `movsx` is a whole dword, and the divide reads its
    halves through an extract. Both are one `idiv r16`.
    """
    extension = definitions.get(high.value)
    if extension is None:
        return False
    if (
        extension.op is ir.Operation.EXTEND
        and extension.name == "cwd"
        and extension.args == (low,)
        and extension.results == (high,)
    ):
        return True
    if (
        extension.kind is not mir.Kind.EXTRACT
        or extension.results != (high,)
        or len(extension.args) != 2
        or extension.loads
        or extension.stores
        or extension.barrier
        or not isinstance(extension.args[0], mir.Held)
        or extension.args[0].width != 4
        or not isinstance(extension.args[1], mir.Const)
        or extension.args[1].n != 16
    ):
        return False
    whole = definitions.get(extension.args[0].value)
    return (
        whole is not None
        and whole.kind is mir.Kind.SIGN_EXTEND
        and whole.args == (low,)
        and whole.results == (extension.args[0],)
        and not whole.loads
        and not whole.stores
        and not whole.barrier
    )

File: src/frontends/bc/raising_division.rs
//! Port of `qbopt/frontend/raising_division.py`: a sign-extended word
//! dividend recognised as one scalar signed division.

use std::collections::BTreeMap;

use num_bigint::BigInt;

use crate::model::ir::Operation;
use crate::model::mir::{Arg, Held, Kind, MirBlock, Op, OpCode, RaisedBody, Value};

pub fn scalar(body: RaisedBody) -> RaisedBody {
    let mut definitions: BTreeMap<Value, &Op> = BTreeMap::new();
    for block in &body.blocks {
        for op in &block.ops {
            for value in &op.defines {
                definitions.insert(*value, op);
            }
        }
    }

    let raised = |op: &Op| -> Op {
        if op.kind != Kind::Div || op.name != "idiv" || op.args.len() != 3 || op.results.len() != 2 {
            return op.clone();
        }
        let (Arg::Held(high), Arg::Held(low), Arg::Held(divisor)) = (&op.args[0], &op.args[1], &op.args[2]) else {
            return op.clone();
        };
        if !op.args.iter().chain(&op.results).all(|arg| matches!(arg, Arg::Held(held) if held.width == 2)) {
            return op.clone();
        }
        if !op.loads.is_empty() || !op.stores.is_empty() || !_extends(high, low, &definitions) {
            return op.clone();
        }
        // `idiv r16` ties dx:ax to the pair it writes; the lowering builds a
        // fresh high from `cwd`, so only a tie to the two halves is known.
        if !op.merges.keys().all(|was| *was == high.value || *was == low.value) {
            return op.clone();
        }
        let remaining = [low.value, divisor.value];
        let mut made = op.clone();
        made.kind = Kind::Divmod;
        made.args = vec![Arg::Held(*low), Arg::Held(*divisor)];
        made.merges = op.merges.iter().filter(|(was, _)| **was != high.value).map(|(was, now)| (*was, *now)).collect();
        made.uses = op.uses.iter().copied().filter(|value| *value != high.value || remaining.contains(value)).collect();
        made
    };

    let blocks: Vec<MirBlock> = body
        .blocks
        .iter()
        .map(|block| MirBlock { ops: block.ops.iter().map(raised).collect(), ..block.clone() })
        .collect();
    body.with_blocks(blocks)
}

/// Whether `high` is the top word of the sign extension of `low`: BC's
/// `cwd`, or a `movsx` dword whose halves the divide reads by extract.
pub fn _extends(high: &Held, low: &Held, definitions: &BTreeMap<Value, &Op>) -> bool {
    let Some(extension) = definitions.get(&high.value) else {
        return false;
    };
    if extension.op == Some(OpCode::Operation(Operation::Extend))
        && extension.name == "cwd"
        && extension.args == [Arg::Held(*low)]
        && extension.results == [Arg::Held(*high)]
    {
        return true;
    }
    if extension.kind != Kind::Extract
        || extension.results != [Arg::Held(*high)]
        || extension.args.len() != 2
        || !extension.loads.is_empty()
        || !extension.stores.is_empty()
        || extension.barrier()
    {
        return false;
    }
    let (Arg::Held(whole_arg), Arg::Const(offset)) = (&extension.args[0], &extension.args[1]) else {
        return false;
    };
    if whole_arg.width != 4 || offset.n != BigInt::from(16) {
        return false;
    }
    let Some(whole) = definitions.get(&whole_arg.value) else {
        return false;
    };
    whole.kind == Kind::SignExtend
        && whole.args == [Arg::Held(*low)]
        && whole.results == [Arg::Held(*whole_arg)]
        && whole.loads.is_empty()
        && whole.stores.is_empty()
        && !whole.barrier()
}

#[cfg(test)]
#[path = "raising_division_tests.rs"]
mod tests;
